crates/controller/src/actions.rs

```diff
@@ -0,0 +1,115 @@
+        pub enum Action {
+            $($action,)*
+            $($building_action,)*
+        }
+
+        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Deserialize, Serialize, Actionlike, Reflect, PartialOrd, Ord)]
+        pub enum MouseAction {
+            $($mouse_action,)*
+        }
+
+        impl DefaultKeybindings for Action {
+            fn default_keybindings() -> InputMap<Self> where Self: Sized {
+                use Action::*;
+                let keybindings = InputMap::<Self>::from(
+                    vec![
+                        $(($action, vec![$($keybind.into()),*])),*,
+                        $(($building_action, vec![$($building_key.into()),*])),*
+                    ].into_iter().collect::<HashMap<Self, Vec<UserInput>>>()
+                );
+                println!("keybindings: {:?}", keybindings);
+                keybindings
+            }
+        }
+
+        impl DefaultKeybindings for MouseAction {
+            fn default_keybindings() -> InputMap<Self> where Self: Sized {
+                use MouseAction::*;
+                let keybindings = InputMap::<Self>::from(
+                    vec![
+                        $(($mouse_action, vec![$($mouse_keybind.into()),*])),*
+                    ].into_iter().collect::<HashMap<Self, Vec<UserInput>>>()
+                );
+                println!("mouse keybindings: {:?}", keybindings);
+                keybindings
+            }
+        }
+
+        impl Action {
+            pub fn get_factory_actions() -> Vec<(Self, de_types::objects::BuildingType)> {
+                use Action::*;
+                use de_types::objects::BuildingType::*;
+
+                vec![$(($building_action, $building_type)),*]
+            }
+        }
+    }
+}
+
+use std::collections::HashMap;
+
+use bevy::app::App;
+use petitset::PetitSet;
+
+make_actions! {
+    // --- general actions ---
+    // keyboard actions
+    Exit, (KeyCode::Escape),
+    SelectAllVisible, (UserInput::chord(vec![KeyCode::ControlLeft, KeyCode::ShiftLeft, KeyCode::A])),
+    SelectAll, (UserInput::chord(vec![KeyCode::ControlLeft, KeyCode::A])),
+    // mouse selections
+    AddToSelection, (
+        UserInput::Chord(PetitSet::from_iter(vec![KeyCode::ControlLeft.into(),MouseButton::Left.into()])),
+        UserInput::Chord(PetitSet::from_iter(vec![KeyCode::ControlRight.into(), MouseButton::Left.into()]))),
+    ReplaceSelection, (MouseButton::Left),
+    // camera controls
+    Up, (KeyCode::W, KeyCode::Up),
+    Down, (KeyCode::S, KeyCode::Down),
+    Left, (KeyCode::A, KeyCode::Left),
+    Right, (KeyCode::D, KeyCode::Right),
+    Pivot, (UserInput::Chord(PetitSet::from_iter(vec![KeyCode::ControlLeft.into(), MouseButton::Middle.into()])));
+    // --- mouse actions (these will trigger the drag logic) ---
+    PrimaryClick, (MouseButton::Left),
+    SecondaryClick, (
+        MouseButton::Right);
+    //  --- building actions ---
```

Comment:
I would merge this to keyboard actions. It can be just `BuildBase, (KeyCode::B)` and so on.